gh-102471, PEP 757: Add PyLong import and export API (#121339)
python/cpython: Doc/c-api/long.rst

   If *op* is compact, as determined by :c:func:`PyUnstable_Long_IsCompact`,
   return its value.

   Otherwise, the return value is undefined.

   .. versionadded:: 3.12

Import/Export API
^^^^^^^^^^^^^^^^^

.. versionadded:: 3.14

.. c:type:: Py_digit

   A single unsigned digit.

   It is usually used in an *array of digits*, such as the
   :c:member:`PyUnstable_LongExport.digits` array.

   Its size depend on the :c:macro:`!PYLONG_BITS_IN_DIGIT` macro:
   see the ``configure`` :option:`--enable-big-digits` option.

   See :c:member:`PyUnstable_Long_LAYOUT.bits_per_digit` for the number of bits per
   digit and :c:member:`PyUnstable_Long_LAYOUT.digit_size` for the size of a digit (in
   bytes).

.. c:struct:: PyUnstable_Long_LAYOUT

   Internal layout of a Python :class:`int` object.

   See also :attr:`sys.int_info` which exposes similar information to Python.

   .. c:member:: uint8_t bits_per_digit;

      Bits per digit.

   .. c:member:: uint8_t digit_size;

      Digit size in bytes.

   .. c:member:: int8_t word_endian;

      Word endian:

      - 1 for most significant byte first (big endian)

>       - 0 for least significant first (little endian)

Reviewer: Same as array_endian.  Also, you are using here PY_LITTLE_ENDIAN.

In short, to have working code on gmpy2 side I did:
```c
        mpz_import(z, layout.ndigits,
                   PyLong_LAYOUT.array_endian ? 1:-1,
                   PyLong_LAYOUT.digit_size,
                   PyLong_LAYOUT.word_endian ? -1:1,
                   PyLong_LAYOUT.digit_size*8 - PyLong_LAYOUT.bits_per_digit,
                   layout.digits);
```

   .. c:member:: int8_t array_endian;

      Array endian:

      - 1 for most significant byte first (big endian)

>       - 0 for least significant first (little endian)

Reviewer: I suggest using here mpz_import convention: "order can be 1 for most significant word first or -1 for least significant first" (c).

> .. c:function:: PyObject* PyUnstable_Long_Import(int negative, size_t ndigits, Py_digit *digits)

Reviewer: Maybe you can add functions/structs without ``PyUnstable_``?  Referenced issue suggests a stable API.

Author: Maybe. I need to think about this.

Reviewer: Either way, you should be consistent in naming.  Test file has remnants with stable API names.

   Create a Python :class:`int` object from an array of digits.

   * Return a Python :class:`int` object on success.
   * Set an exception and return ``NULL`` on error.

   *negative* is ``1`` if the number is negative, or ``0`` otherwise.

   *ndigits* is the number of digits in the *digits* array.

   *digits* is an array of unsigned digits.

   See :c:struct:`PyUnstable_Long_LAYOUT` for the internal layout of an integer.

> .. c:struct:: PyUnstable_LongExport

Reviewer: This clash with the function name.

```suggestion
.. c:struct:: PyUnstable_LongDigitsArray
```

   A Python :class:`int` object exported as an array of digits.

   See :c:struct:`PyUnstable_Long_LAYOUT` for the internal layout of an integer.

   .. c:member:: PyLongObject *obj

      Strong reference to the Python :class:`int` object.

   .. c:member:: int negative

      1 if the number is negative, 0 otherwise.

   .. c:member:: size_t ndigits

      Number of digits in :c:member:`digits` array.

   .. c:member:: Py_digit *digits

      Array of unsigned digits.

.. c:function:: int PyUnstable_Long_Export(PyObject *obj, PyUnstable_LongExport *export)

   Export a Python :class:`int` object as an array of digits.

   * Set *\*export* and return 0 on success.
   * Set an exception and return -1 on error.

> 

Reviewer: ```suggestion

   This function always succeeds if *obj* is a :c:type:`PyLongObject` or its subtype.

```

   see the ``configure`` :option:`--enable-big-digits` option.

   See :c:member:`PyLong_LAYOUT.bits_per_digit` for the number of bits per
   digit and :c:member:`PyLong_LAYOUT.digit_size` for the size of a digit (in
   bytes).

.. c:struct:: PyLong_LAYOUT

   Layout of an array of digits, used by Python :class:`int` object.

   See also :attr:`sys.int_info` which exposes similar information to Python.

   .. c:member:: uint8_t bits_per_digit

      Bits per digit.

   .. c:member:: uint8_t digit_size

      Digit size in bytes.

   .. c:member:: int8_t word_endian

      Word endian:

      - ``1`` for most significant word first (big endian)
      - ``-1`` for least significant first (little endian)

   .. c:member:: int8_t array_endian

      Array endian:

      - ``1`` for most significant byte first (big endian)
      - ``-1`` for least significant first (little endian)

.. c:struct:: PyLong_DigitArray

   A Python :class:`int` object exported as an array of digits.

   See :c:struct:`PyLong_LAYOUT` for the :c:member:`digits` layout.

   .. c:member:: PyObject *obj

      Strong reference to the Python :class:`int` object.

>    .. c:member:: int negative
> 
>       1 if the number is negative, 0 otherwise.
> 
>    .. c:member:: size_t ndigits
> 
>       Number of digits in :c:member:`digits` array.

Reviewer: The negative field is a convenience.  We have API to access sign and I can imagine scenario when someone will be interested just in the absolute value.  So, you could shrink structure by removing this field:
```suggestion
   .. c:member:: size_t ndigits

      Number of digits in :c:member:`digits` array.
```

I'm less certain about ndigits field (which should be of Py_ssize_t, probably).  Maybe it does make sense to expose _PyLong_DigitCount() with API like PyLong_GetSign()?

Author: I prefer to keep negative member, it makes the whole structure complete.

Reviewer: Actually, arbitrary precision integer just can't be represented as "array of digits").  That works only for natural numbers.

.. c:function:: int PyLong_Export(PyObject *obj, PyLong_DigitArray *array)

   Export a Python :class:`int` object as an array of digits.

   On success, set *\*array* and return 0.
   On error, set an exception and return -1.

   This function always succeeds if *obj* is a Python :class:`int` object or a
   subclass.

   :c:func:`PyLong_ReleaseExport` must be called once done with using
   *export*.

.. c:function:: void PyLong_ReleaseExport(PyLong_DigitArray *array)

   Release the export *array* created by :c:func:`PyLong_Export`.

PyLongWriter API
^^^^^^^^^^^^^^^^

The :c:type:`PyLongWriter` API can be used to import an integer.

.. versionadded:: 3.14

.. c:struct:: PyLongWriter

   A Python :class:`int` writer instance.

   The instance must be destroyed by :c:func:`PyLongWriter_Finish`.

.. c:function:: PyLongWriter* PyLongWriter_Create(int negative, size_t ndigits, Py_digit **digits)

   Create a :c:type:`PyLongWriter`.

   On success, set *\*digits* and return a writer.
   On error, set an exception and return ``NULL``.

   *negative* is ``1`` if the number is negative, or ``0`` otherwise.

   *ndigits* is the number of digits in the *digits* array.

>    The caller must initialize the array of digits *digits* and then call
>    :c:func:`PyLongWriter_Finish` to get a Python :class:`int`. Digits must be
>    in the range [``0``; ``PyLong_BASE - 1``].

Reviewer: ```suggestion
   The caller must initialize the array of digits *digits* and then call
   :c:func:`PyLongWriter_Finish` to get a Python :class:`int`. Digits must be
   in the range [``0``; ``PyLong_BASE - 1``].  Unused digits must be set to ``0``.
```

.. c:function:: Py_ssize_t PyUnstable_Long_CompactValue(const PyLongObject* op)

   If *op* is compact, as determined by :c:func:`PyUnstable_Long_IsCompact`,
   return its value.

   Otherwise, the return value is undefined.

   .. versionadded:: 3.12

Import/Export API
^^^^^^^^^^^^^^^^^

.. versionadded:: 3.14

.. c:type:: Py_digit

   A single unsigned digit in the range [``0``; ``PyLong_BASE - 1``].

   It is usually used in an *array of digits*, such as the
   :c:member:`PyLong_DigitArray.digits` array.

   Its size depend on the :c:macro:`!PYLONG_BITS_IN_DIGIT` macro:
   see the ``configure`` :option:`--enable-big-digits` option.

   See :c:member:`PyLong_LAYOUT.bits_per_digit` for the number of bits per
   digit and :c:member:`PyLong_LAYOUT.digit_size` for the size of a digit (in
   bytes).

.. c:struct:: PyLong_LAYOUT

   Layout of an array of digits, used by Python :class:`int` object.

   See also :attr:`sys.int_info` which exposes similar information to Python.

   .. c:member:: uint8_t bits_per_digit

      Bits per digit.

   .. c:member:: uint8_t digit_size

      Digit size in bytes.

   .. c:member:: int8_t word_endian

      Word endian:

>       - ``1`` for most significant byte first (big endian)

Reviewer: ```suggestion
      - ``1`` for most significant word first (big endian)
```

      - ``-1`` for least significant first (little endian)

   .. c:member:: int8_t array_endian

      Array endian:

      - ``1`` for most significant byte first (big endian)
      - ``-1`` for least significant first (little endian)

.. c:function:: PyObject* PyLong_Import(int negative, size_t ndigits, Py_digit *digits)

   Create a Python :class:`int` object from an array of digits.

   On success, return a Python :class:`int` object.
   On error, set an exception and return ``NULL``.

   *negative* is ``1`` if the number is negative, or ``0`` otherwise.

   *ndigits* is the number of digits in the *digits* array.

>    *digits* is an array of unsigned digits. Digits must be in the range
>    [``0``; ``PyLong_BASE - 1``].

Reviewer: This macro is undocumented.

BTW, commentary in longintrepr.h says: "Type 'digit' should be able to hold 2*PyLong_BASE-1":
```pycon
>>> _testcapi.pylong_export(2*(1<<sys.int_info.bits_per_digit)-1)
(0, [1073741823, 1])
>>> _testcapi.pylong_export(2*(1<<sys.int_info.bits_per_digit))
(0, [0, 2])
>>> _testcapi.pylong_import(0, [0, 2])
2147483648
>>> 2*(1<<sys.int_info.bits_per_digit)
2147483648
```
```suggestion
   *digits* is an array of unsigned digits. Digits must be in the range
   [``0``; ``2**(PyLong_SHIFT+1) - 1``], where :c:macro:`!PyLong_SHIFT`
   is equal to :c:member:`PyLong_LAYOUT.bits_per_digit`.
```

Author: > BTW, commentary in longintrepr.h says: "Type 'digit' should be able to hold 2*PyLong_BASE-1"

The `digit` type is used for computation, it should support a range larger than Python int digit range. Python int digits must be in the range [0; PyLong_BASE-1].

   See :c:struct:`PyLong_LAYOUT` for the layout of an array of digits.

   See also the :ref:`PyLongWriter API <pylongwriter>`.

.. c:struct:: PyLong_DigitArray

   A Python :class:`int` object exported as an array of digits.

   See :c:struct:`PyLong_LAYOUT` for the :c:member:`digits` layout.

   .. c:member:: PyLongObject *obj

      Strong reference to the Python :class:`int` object.

   .. c:member:: int negative

      1 if the number is negative, 0 otherwise.

   .. c:member:: size_t ndigits

      Number of digits in :c:member:`digits` array.

   .. c:member:: Py_digit *digits

      Array of unsigned digits.

> .. c:function:: int PyLong_Export(PyObject *obj, PyLong_DigitArray *array)

Reviewer: Maybe
```suggestion
.. c:function:: PyLong_DigitArray* PyLong_Export(PyObject *obj)
```
?

Or in this way you are trying to emphasize that *array* is read-only?

Author: I prefer to allocate `PyLong_DigitArray` on the stack memory and pass `&array`. It avoids one memory allocation on the heap.

Also, I like the return 0 for success and -1 for error: https://devguide.python.org/developer-workflow/c-api/index.html#guidelines-for-expanding-changing-the-public-api

Reviewer: I'll try to benchmark this.

With a different signature you could do declaration & initialization of PyLong_DigitArray in one line.

> Also, I like the return 0 for success and -1 for error

I don't think this guide ban using NULL to signal error (which nobody will actually check;)).

   Export a Python :class:`int` object as an array of digits.

   On success, set *\*array* and return 0.
   On error, set an exception and return -1.

   This function always succeeds if *obj* is a Python :class:`int` object or a
   subclass.

   :c:func:`PyLong_ReleaseExport` must be called once done with using
   *export*.

.. c:function:: void PyLong_ReleaseExport(PyLong_DigitArray *array)

   Release the export *array* created by :c:func:`PyLong_Export`.

.. _pylongwriter:

PyLongWriter API
^^^^^^^^^^^^^^^^

The :c:func:`PyLong_Import` function can be used to create a Python
:class:`int` from an array of digits. When converting an integer from a format
to a Python :class:`int`, creating a temporary array of digits can be
inefficient. In this case, the :c:type:`PyLongWriter` API can be used for even

> more efficient import.

Reviewer: I doubt this will help users.  I.e. it's not clear when PyLong_Import() may be inefficient and we can't provide this information.

Lets keep one way to do things.  PyLong_Import() is a redundant function, it should be removed; and something like PyLongWriter_Create/Finish - simple enough interface for writing.

Author: Ok, I removed PyLong_Import().

.. versionadded:: 3.14

.. c:struct:: PyLongWriter

   A Python :class:`int` writer instance.

   The instance must be destroyed by :c:func:`PyLongWriter_Finish`.

> .. c:function:: PyLongWriter* PyLongWriter_Create(int negative, size_t ndigits, Py_digit **digits)

Reviewer: This function has two return values (writer and digits).  Is this a good price to hide PyLongWriter structure details?  May we want for PyLongWriter something different than PyLong_DigitArray?

Author: There is an optimization: there is no real PyLongWriter structure in fact, just a cast between pointer types. It's not uncommon for functions to return more than one value.

Structures can also make the maintenance and evolution of this API more complicated. Programming languages which cannot use C header files have to reimplement the structure, it's more work for them.

   Create a :c:type:`PyLongWriter`.

   On success, set *\*digits* and return a writer.
   On error, set an exception and return ``NULL``.

   *negative* is ``1`` if the number is negative, or ``0`` otherwise.

   *ndigits* is the number of digits in the *digits* array.

   The caller must initialize the array of digits *digits* and then call
   :c:func:`PyLongWriter_Finish` to get a Python :class:`int`. Digits must be

>    in the range [``0``; ``PyLong_BASE - 1``].

Reviewer: ```suggestion
   in the range [``0``; ``2*PyLong_BASE - 1``].
```

   See :c:struct:`PyLong_LAYOUT` for the layout of an array of digits.

   See also the :c:func:`PyLong_Import` function.

.. c:function:: PyObject* PyLongWriter_Finish(PyLongWriter *writer)

   Finish a :c:type:`PyLongWriter` created by :c:func:`PyLongWriter_Create`.

   On success, return a Python :class:`int` object.
   On error, set an exception and return ``NULL``.

> Example with a single digit::

Reviewer: It might worth to keep this example while discussing the API but, honestly, I find documentation clear enough without.

Author: Ok, I removed the example.

   Exactly what values are considered compact is an implementation detail
   and is subject to change.

   .. versionadded:: 3.12

.. c:function:: Py_ssize_t PyUnstable_Long_CompactValue(const PyLongObject* op)

   If *op* is compact, as determined by :c:func:`PyUnstable_Long_IsCompact`,
   return its value.

   Otherwise, the return value is undefined.

   .. versionadded:: 3.12

Export API
^^^^^^^^^^

.. versionadded:: 3.14

.. c:struct:: PyLongLayout

   Layout of an array of digits, used by Python :class:`int` object.

   Use :c:func:`PyLong_GetNativeLayout` to get the native layout of Python
   :class:`int` objects.

   See also :attr:`sys.int_info` which exposes similar information to Python.

   .. c:member:: uint8_t bits_per_digit

      Bits per digit.

   .. c:member:: uint8_t digit_size

      Digit size in bytes.

   .. c:member:: int8_t digits_order

>       Word endian:

Reviewer 2: What is "word" here? This is a bit confusing.

>       - ``1`` for most significant word first (big endian)

Reviewer 2: ```suggestion
      - ``1`` for most significant digit first (big endian)
```

      - ``-1`` for least significant first (little endian)

>    .. c:member:: int8_t endian

Reviewer 2: ```suggestion
   .. c:member:: int8_t endianness
```

Author: If we rename the member, I would prefer to do it first in the PEP (the reference working document). https://github.com/python/peps/pull/3970 PEP change is being discussed, we can change the member name once this PEP change is merged.

Author: Done in https://github.com/python/peps/commit/9475fa0aa497fc373d2cf9ed4f94c952e6368326.

>       Array endian:

Reviewer 2: Why "Array"? 
```suggestion
      Digit endianness:
```

      - ``1`` for most significant byte first (big endian)
      - ``-1`` for least significant first (little endian)

.. c:function:: const PyLongLayout* PyLong_GetNativeLayout(void)

   Get the native layout of Python :class:`int` objects.

   See the :c:struct:`PyLongLayout` structure.

.. c:struct:: PyLongExport

   Export of a Python :class:`int` object.

   There are two cases:

   * If :c:member:`digits` is ``NULL``, only use the :c:member:`value` member.
     Calling :c:func:`PyLong_FreeExport` is optional in this case.
   * If :c:member:`digits` is not ``NULL``, use :c:member:`negative`,
     :c:member:`ndigits` and :c:member:`digits` members.
     Calling :c:func:`PyLong_FreeExport` is mandatory in this case.

   .. c:member:: int64_t value

>       Integer value if :c:member:`digits` is ``NULL``.

Reviewer 2: ### 
```suggestion
      The native integer value of the exported :class:`int` object.
      Only valid if :c:member:`digits` is ``NULL``.
```

   .. c:member:: uint8_t negative

>       1 if the number is negative, 0 otherwise.

Reviewer 2: ```suggestion
      1 if the number is negative, 0 otherwise.
      Only valid if :c:member:`digits` is not ``NULL``.
```

   .. c:member:: Py_ssize_t ndigits

>       Number of digits in :c:member:`digits` array.

Reviewer 2: ```suggestion
      Number of digits in :c:member:`digits` array.
      Only valid if :c:member:`digits` is not ``NULL``.
```

   Export a Python :class:`int` object.

   On success, set *\*export_long* and return 0.
   On error, set an exception and return -1.

   This function always succeeds if *obj* is a Python :class:`int` object or a
   subclass.

   If *export_long.digits* is not ``NULL``, :c:func:`PyLong_FreeExport` must be
   called when the export is no longer needed.

.. c:function:: void PyLong_FreeExport(PyLongExport *export_long)

   Release the export *export_long* created by :c:func:`PyLong_Export`.

PyLongWriter API
^^^^^^^^^^^^^^^^

The :c:type:`PyLongWriter` API can be used to import an integer.

.. versionadded:: 3.14

.. c:struct:: PyLongWriter

   A Python :class:`int` writer instance.

   The instance must be destroyed by :c:func:`PyLongWriter_Finish`.

.. c:function:: PyLongWriter* PyLongWriter_Create(int negative, Py_ssize_t ndigits, void **digits)

   Create a :c:type:`PyLongWriter`.

   On success, set *\*digits* and return a writer.
   On error, set an exception and return ``NULL``.

   *negative* is ``1`` if the number is negative, or ``0`` otherwise.

   *ndigits* is the number of digits in the *digits* array. It must be
   greater than or equal to 0.

> 

Reviewer 2: ```suggestion
   *digits* must point to a caller-allocated memory area large enough to host
   *ndigits* digits of the size given by :c:member:`PyLongLayout.digit_size`.
   The *digits* pointer must remain valid until either :c:func:`PyLongWriter_Finish`
   or :c:func:`PyLongWriter_Discard` is called, after which the caller is free to
   deallocate it if needed.

```

Author: That's wrong: the *digits* array is allocated by PyLongWriter_Create().

Reviewer 2: Oh! Well, it seems it must better be explained then :-))

Author: Sure, there is room for enhancement on this doc :-)

   See :c:member:`PyLongLayout.bits_per_digit` for the number of bits per
   digit and :c:member:`PyLongLayout.digit_size` for the size of a digit (in
   bytes).

.. c:struct:: PyLongLayout

   Layout of an array of digits, used by Python :class:`int` object.

   Use :c:func:`PyLong_GetNativeLayout` to get the native layout of Python
   :class:`int` objects.

   See also :attr:`sys.int_info` which exposes similar information to Python.

   .. c:member:: uint8_t bits_per_digit

      Bits per digit.

   .. c:member:: uint8_t digit_size

      Digit size in bytes.

   .. c:member:: int8_t digits_order

      Word endian:

      - ``1`` for most significant word first (big endian)
      - ``-1`` for least significant first (little endian)

   .. c:member:: int8_t endian

      Array endian:

      - ``1`` for most significant byte first (big endian)
      - ``-1`` for least significant first (little endian)

.. c:function:: const PyLongLayout* PyLong_GetNativeLayout(void)

   Get the native layout of Python :class:`int` objects.

   See the :c:struct:`PyLongLayout` structure.

.. c:struct:: PyLong_DigitArray

   A Python :class:`int` object exported as an array of digits.

>    .. c:member:: PyObject *obj

Reviewer 3: Is there any use for this by the consumer? Seems like the purpose of this field is to keep alive the corresponding Python object, but if `PyLong_AsDigitArray` choses to create a fresh digits array for some reason (different internal representation), then this would be probably `NULL`, right?

1) I fear people will start relying on this being the object unconditionally and will just use it as a convenience (I do not need to pass around the pointer to the object, because I have it in `PyLong_DigitArray`), 
2) if there is no intended use for this on the consumer side, I would defensively hide it as much as possible, 
3) I would make it opaque `void* _reserved`, so the runtime may chose to use it for something else that is useful in the `PyLong_FreeDigitArray` function (pointer to some destructor function, additional data somehow needed to free the array, ...). 

Moreover, (at least) in debug builds I propose to transform the PyObject* pointer, e.g., do `+/-1`, so that accessing the `void*` as `PyObject*` in the extension will segfault (I actually think that the miniscule perf penalty vs. preventing API misuse would make it worth be enabled always).

Author: Good idea. I wrote https://github.com/python/peps/pull/3966 to make the member private: replace `PyObject *obj` with `Py_uintptr_t _reserved`.

      1 if the number is negative, 0 otherwise.

   .. c:member:: Py_ssize_t ndigits

      Number of digits in :c:member:`digits` array.

   .. c:member:: const void *digits

      Read-only array of unsigned digits.

>    .. c:member:: Py_uintptr_t reserved

Reviewer: Should we document internal fields at all?

Author: I removed the documentation of this member.

      Read-only array of unsigned digits.

.. c:function:: int PyLong_AsDigitArray(PyObject *obj, PyLong_DigitArray *array)

   Export a Python :class:`int` object as an array of digits.

   On success, set *\*array* and return 0.
   On error, set an exception and return -1.

   This function always succeeds if *obj* is a Python :class:`int` object or a

>    subclass.

Reviewer: @vstinner, after some thinking I believe we should drop this contract, if we take into account future changes in internals of CPython's integers.  If single layout view will be invalid - this function, probably, will allocate temporary buffers.  That might fail.

Can we offer a different contract instead, something like this: "This function always succeeds if *obj* is a Python `int` object or a subclass *and* it's value can't be converted to C long."?  In this case users have a clear hint: "try something like PyLong_AsLongAndOverflow and if it fails - fallback to this function".  If not, I think this is a severe issue with that part of API.

Reviewer 2: No fallback to `PyLong_AsLongAndOverflow` should be needed with this API, thanks to the embedded `int64_t value`.

   Exactly what values are considered compact is an implementation detail
   and is subject to change.

   .. versionadded:: 3.12

.. c:function:: Py_ssize_t PyUnstable_Long_CompactValue(const PyLongObject* op)

   If *op* is compact, as determined by :c:func:`PyUnstable_Long_IsCompact`,
   return its value.

   Otherwise, the return value is undefined.

   .. versionadded:: 3.12

Export API
^^^^^^^^^^

.. versionadded:: 3.14

.. c:struct:: PyLongLayout

   Layout of an array of digits, used by Python :class:`int` object.

   Use :c:func:`PyLong_GetNativeLayout` to get the native layout of Python
   :class:`int` objects.

>    See also :attr:`sys.int_info` which exposes similar information to Python.

Reviewer: ```suggestion
   See also :data:`sys.int_info` which exposes similar information to Python.
```

   .. versionadded:: 3.12

Export API
^^^^^^^^^^

.. versionadded:: 3.14

.. c:struct:: PyLongLayout

   Layout of an array of "digits" ("limbs" in the GMP terminology), used to
   represent absolute value for arbitrary precision integers.

   Use :c:func:`PyLong_GetNativeLayout` to get the native layout of Python
   :class:`int` objects, used internally for integers with "big enough"
   absolute value.

>    See also :data:`sys.int_info` which exposes similar information to Python.

Reviewer 4: ```suggestion
   See also :data:`sys.int_info` which exposes similar information in Python.
```

Technically "to" is correct, but it's ambiguous (is the information _exposed_ to, or _similar_ to?). "In" Python is clearer, even though it's not quite right.

   .. c:member:: uint8_t bits_per_digit

>       Bits per digit.

Reviewer 4: Could you add an example of how to interpret this? Something like: "Bits per digit. For example, a 15 bit digit means that bits 0-14 contain meaningful information."

Similar for `digit_size`: "Digit size in bytes. For example, a 15 bit digit will require at least 2 bytes."

You might have better information to communicate here, my aim is to save people (i.e. me) from having to go read the PEP or the source code to figure out how to use these correctly.

Author: Done, I added your suggestion.

   * If :c:member:`digits` is ``NULL``, only use the :c:member:`value` member.
     Calling :c:func:`PyLong_FreeExport` is optional in this case.
   * If :c:member:`digits` is not ``NULL``, use :c:member:`negative`,
     :c:member:`ndigits` and :c:member:`digits` members.
     Calling :c:func:`PyLong_FreeExport` is mandatory in this case.

   .. c:member:: int64_t value

      The native integer value of the exported :class:`int` object.
      Only valid if :c:member:`digits` is ``NULL``.

   .. c:member:: uint8_t negative

      ``1`` if the number is negative, ``0`` otherwise.
      Only valid if :c:member:`digits` is not ``NULL``.

   .. c:member:: Py_ssize_t ndigits

      Number of digits in :c:member:`digits` array.
      Only valid if :c:member:`digits` is not ``NULL``.

   .. c:member:: const void *digits

      Read-only array of unsigned digits. Can be ``NULL``.

.. c:function:: int PyLong_Export(PyObject *obj, PyLongExport *export_long)

   Export a Python :class:`int` object.

   On success, set *\*export_long* and return ``0``.
   On error, set an exception and return ``-1``.

   If *export_long->digits* is not ``NULL``, :c:func:`PyLong_FreeExport` must

>    be called when the export is no longer needed.

Reviewer 4: What happens if we call `PyLong_FreeExport` when `digits` _is_ `NULL`? Is that okay? (It should be okay - much easier to balance alloc/free calls when you can always call them. It can remove an entire class of logic bugs.)

Perhaps we should move this bit to the `FreeExport` doc and say "If `digits` was returned as `NULL`, calling this function is optional"?

Author: Calling PyLong_FreeExport() is always safe. It can be called if digits is not NULL, or if digits is NULL.

*Implementation detail: it can even be called twice.*

Reviewer 5: Do you want to mention that `export_long` should not be NULL (or else it crashes?)

Reviewer 4: Sorry, I meant you should write that in the docs, not tell me in a comment.

Author: Ok, I changed the doc to:

```
   If *export_long->digits* is not ``NULL``, :c:func:`PyLong_FreeExport` must
   be called when the export is no longer needed. Otherwise, calling
   :c:func:`PyLong_FreeExport` is optional.
```

(I added "Otherwise, ...".)

   .. versionadded:: 3.12

Export API
^^^^^^^^^^

> .. versionadded:: 3.14

Reviewer 5: ```suggestion
.. versionadded:: next
```

.. c:struct:: PyLongLayout

   Layout of an array of "digits" ("limbs" in the GMP terminology), used to
   represent absolute value for arbitrary precision integers.

   Use :c:func:`PyLong_GetNativeLayout` to get the native layout of Python
   :class:`int` objects, used internally for integers with "big enough"
   absolute value.

   See also :data:`sys.int_info` which exposes similar information in Python.

   .. c:member:: uint8_t bits_per_digit

      Bits per digit. For example, a 15 bit digit means that bits 0-14 contain
      meaningful information.

   .. c:member:: uint8_t digit_size

      Digit size in bytes. For example, a 15 bit digit will require at least 2
      bytes.

   .. c:member:: int8_t digits_order

      Digits order:

      - ``1`` for most significant digit first
      - ``-1`` for least significant digit first

   .. c:member:: int8_t digit_endianness

      Digit endianness:

      - ``1`` for most significant byte first (big endian)

>       - ``-1`` for least significant first (little endian)

Reviewer 5: ```suggestion
      - ``-1`` for least significant byte first (little endian)
```

     Calling :c:func:`PyLong_FreeExport` is optional in this case.
   * If :c:member:`digits` is not ``NULL``, use :c:member:`negative`,
     :c:member:`ndigits` and :c:member:`digits` members.
     Calling :c:func:`PyLong_FreeExport` is mandatory in this case.

   .. c:member:: int64_t value

      The native integer value of the exported :class:`int` object.
      Only valid if :c:member:`digits` is ``NULL``.

   .. c:member:: uint8_t negative

      ``1`` if the number is negative, ``0`` otherwise.
      Only valid if :c:member:`digits` is not ``NULL``.

   .. c:member:: Py_ssize_t ndigits

      Number of digits in :c:member:`digits` array.
      Only valid if :c:member:`digits` is not ``NULL``.

   .. c:member:: const void *digits

      Read-only array of unsigned digits. Can be ``NULL``.

.. c:function:: int PyLong_Export(PyObject *obj, PyLongExport *export_long)

   Export a Python :class:`int` object.

   On success, set *\*export_long* and return ``0``.
   On error, set an exception and return ``-1``.

   If *export_long->digits* is not ``NULL``, :c:func:`PyLong_FreeExport` must
   be called when the export is no longer needed.

> 

Reviewer: ```suggestion

    .. impl-detail::

        This function always succeeds if *obj* is a Python :class:`int` object
        or a subclass.

```
Lets see if we can restore this in a that way.  It might be helpful for e.g. Sage, which doesn't support PyPy.

Author: I would prefer to not add this note. It was controversial during PEP 757 design.

@serhiy-storchaka @encukou: What do you think? Would you be ok to declare that the PyLong_Export() function cannot fail if the argument is a Python int?

Reviewer: > It was controversial during PEP 757 design.

It was proposed unconditionally, not as CPython's implementation detail.

Reviewer 6: I'm fine with it, as an implementation detail.

Reviewer: @serhiy-storchaka ?

.. c:function:: void PyLong_FreeExport(PyLongExport *export_long)

   Release the export *export_long* created by :c:func:`PyLong_Export`.

PyLongWriter API
^^^^^^^^^^^^^^^^

The :c:type:`PyLongWriter` API can be used to import an integer.

> .. versionadded:: 3.14

Reviewer 5: ```suggestion
.. versionadded:: next
```

.. c:struct:: PyLongWriter

   A Python :class:`int` writer instance.

   The instance must be destroyed by :c:func:`PyLongWriter_Finish` or
   :c:func:`PyLongWriter_Discard`.

.. c:function:: PyLongWriter* PyLongWriter_Create(int negative, Py_ssize_t ndigits, void **digits)

   Create a :c:type:`PyLongWriter`.

   On success, allocate *\*digits* and return a writer.

>    On error, set an exception and return ``NULL``.

Reviewer 5: Does it set `digits` to NULL in case of an error? (apparently not, but do you want to?)

Reviewer: I don't think that should be specified.

Author: I modified the implementation to set `*digits` to NULL on error, but I agree with @skirpichev, I don't think that it should be specified.

   *negative* is ``1`` if the number is negative, or ``0`` otherwise.

   *ndigits* is the number of digits in the *digits* array. It must be
   greater than 0.

   *digits* must not be NULL.

   The caller can either initialize the array of digits *digits* and then
   either call :c:func:`PyLongWriter_Finish` to get a Python :class:`int` or
   :c:func:`PyLongWriter_Discard` to destroy the writer instance.  Digits must
   be in the range [``0``; ``(1 << bits_per_digit) - 1``]  (where the
   :c:struct:`~PyLongLayout.bits_per_digit` is the number of bits per digit).

>    The unused most-significant digits must be set to ``0``.

Reviewer 5: ```suggestion
   The unused most significant digits must be set to ``0``.
```

.. c:function:: PyObject* PyLongWriter_Finish(PyLongWriter *writer)

   Finish a :c:type:`PyLongWriter` created by :c:func:`PyLongWriter_Create`.

   On success, return a Python :class:`int` object.
   On error, set an exception and return ``NULL``.

   The function takes care of normalizing the digits and converts the object
   to a compact integer if needed.

   The writer instance is invalid after the call.

.. c:function:: void PyLongWriter_Discard(PyLongWriter *writer)

>    Discard a :c:type:`PyLongWriter` created by :c:func:`PyLongWriter_Create`.

Reviewer 5: Does it take care of checking for `writer == NULL`? For unicode writers, docs say "If writer is NULL, no operation is performed." (apparently, not but do you want to?)

Reviewer: I think it's better to omit this for such API.  The behavior is covered by general note in docs: https://docs.python.org/3/c-api/concrete.html

Reviewer 5: Yes, but there are many pages in the C API and it doesn't hurt to recall stuff like this. A general note that is far from where it's being used is generally not helpful (especially if people are coming to the link from the What's New page)

Author: writer must not be NULL.

   :class:`int` objects, used internally for integers with "big enough"
   absolute value.

   See also :data:`sys.int_info` which exposes similar information in Python.

   .. c:member:: uint8_t bits_per_digit

      Bits per digit. For example, a 15 bit digit means that bits 0-14 contain
      meaningful information.

   .. c:member:: uint8_t digit_size

      Digit size in bytes. For example, a 15 bit digit will require at least 2
      bytes.

   .. c:member:: int8_t digits_order

      Digits order:

      - ``1`` for most significant digit first
      - ``-1`` for least significant digit first

   .. c:member:: int8_t digit_endianness

      Digit endianness:

      - ``1`` for most significant byte first (big endian)
      - ``-1`` for least significant byte first (little endian)

.. c:function:: const PyLongLayout* PyLong_GetNativeLayout(void)

   Get the native layout of Python :class:`int` objects.

   See the :c:struct:`PyLongLayout` structure.

   The function must not be called before Python initialization nor after
   Python finalization. The returned layout is valid until Python is
   finalized. The layout is the same for all Python sub-interpreters and

>    so it can be cached.

Reviewer 6: I prefer being explicit with words like “global” and “all” :)
```suggestion
   finalized. The layout is the same for all Python sub-interpreters
   in a process, and so it can be cached.
```

.. c:struct:: PyLongExport

   Export of a Python :class:`int` object.

   There are two cases:

   * If :c:member:`digits` is ``NULL``, only use the :c:member:`value` member.
     Calling :c:func:`PyLong_FreeExport` is optional in this case.
   * If :c:member:`digits` is not ``NULL``, use :c:member:`negative`,
     :c:member:`ndigits` and :c:member:`digits` members.
     Calling :c:func:`PyLong_FreeExport` is mandatory in this case.

   .. c:member:: int64_t value

      The native integer value of the exported :class:`int` object.
      Only valid if :c:member:`digits` is ``NULL``.

   .. c:member:: uint8_t negative

      ``1`` if the number is negative, ``0`` otherwise.
      Only valid if :c:member:`digits` is not ``NULL``.

   .. c:member:: Py_ssize_t ndigits

      Number of digits in :c:member:`digits` array.
      Only valid if :c:member:`digits` is not ``NULL``.

   .. c:member:: const void *digits

      Read-only array of unsigned digits. Can be ``NULL``.

.. c:function:: int PyLong_Export(PyObject *obj, PyLongExport *export_long)

   Export a Python :class:`int` object.

>    *export_long* must not be ``NULL``.

Reviewer 6: Consider making the following clearer, so the reader doesn't need to count the `*`s in `PyLongExport *export_long`:
```suggestion
   *export_long* must point to a :c:struct:`PyLongExport` structure
   allocated by the caller. It must not be ``NULL``.
```

>    On success, set *\*export_long* and return ``0``.

Reviewer 6: Here, *set* brings to mind writing a single integer.
```suggestion
   On success, fill in *\*export_long* and return ``0``.
```

   Release the export *export_long* created by :c:func:`PyLong_Export`.

PyLongWriter API
^^^^^^^^^^^^^^^^

The :c:type:`PyLongWriter` API can be used to import an integer.

.. versionadded:: next

.. c:struct:: PyLongWriter

   A Python :class:`int` writer instance.

   The instance must be destroyed by :c:func:`PyLongWriter_Finish` or
   :c:func:`PyLongWriter_Discard`.

.. c:function:: PyLongWriter* PyLongWriter_Create(int negative, Py_ssize_t ndigits, void **digits)

   Create a :c:type:`PyLongWriter`.

   On success, allocate *\*digits* and return a writer.
   On error, set an exception and return ``NULL``.

   *negative* is ``1`` if the number is negative, or ``0`` otherwise.

   *ndigits* is the number of digits in the *digits* array. It must be
   greater than 0.

   *digits* must not be NULL.

   The caller can either initialize the array of digits *digits* and then
   either call :c:func:`PyLongWriter_Finish` to get a Python :class:`int` or
   :c:func:`PyLongWriter_Discard` to destroy the writer instance.  Digits must
   be in the range [``0``; ``(1 << bits_per_digit) - 1``]  (where the
   :c:struct:`~PyLongLayout.bits_per_digit` is the number of bits per digit).

>    The unused most significant digits must be set to ``0``.

Reviewer 6: ```suggestion
   After a successful call to this function,
   the caller should fill in the array of digits *digits* and then
   call :c:func:`PyLongWriter_Finish` to get a Python :class:`int`.
   The layout of *digits* is described by :c:func:`PyLong_GetNativeLayout`.
   Digits must
   be in the range [``0``; ``(1 << bits_per_digit) - 1``]  (where the
   :c:struct:`~PyLongLayout.bits_per_digit` is the number of bits per digit).
   Any unused most significant digits must be set to ``0``.

   Alternately, call :c:func:`PyLongWriter_Discard` to destroy the
   writer instance without creating an :class:`~int` object.
```

.. c:function:: PyObject* PyLongWriter_Finish(PyLongWriter *writer)

   Finish a :c:type:`PyLongWriter` created by :c:func:`PyLongWriter_Create`.

   On success, return a Python :class:`int` object.
   On error, set an exception and return ``NULL``.

   The function takes care of normalizing the digits and converts the object
   to a compact integer if needed.

>    The writer instance is invalid after the call.

Reviewer 6: ```suggestion
   The writer instance and the *digits* array are invalid after the call.
```

.. c:function:: void PyLongWriter_Discard(PyLongWriter *writer)

   Discard a :c:type:`PyLongWriter` created by :c:func:`PyLongWriter_Create`.

   *writer* must not be ``NULL``.

>    The writer instance is invalid after the call.

Reviewer 6: ```suggestion
   The writer instance and the *digits* array are invalid after the call.
```

   *negative* is ``1`` if the number is negative, or ``0`` otherwise.

   *ndigits* is the number of digits in the *digits* array. It must be

>    positive.

Reviewer 5: To avoid any ambiguities, I would say "nonnegative" or "greater than or equal to 0".

Author: Ok, done
